### PR TITLE
Docker-compose support for MTConnect-SmartAdapter and ODS with scripts update

### PR DESCRIPTION
**Pull request by emprarthanak**
Jira Id: EP4US1

- Updated docker-compose.yml script to support MTConnect-SmartAdapter and ODS
- Updated ssInstall, ssUpgrade, ssClean and ssStatus scripts 
- Added Functional Specification Document for SmartSaw Upgrade and Docker-compose

### CODE REVIEW

**Comment by skibum1869** [comment]
We will run a test this week or early next week on a saw being built.

File: ssUpgrade.sh
Comment (skibum1869) [inline ssUpgrade.sh:1]: in `cp -r ./ods/config/. /etc/ods/config` the `./ods/config/.` is not a valid folder. the ods folder link is not correct

**Comment by skibum1869** [comment]
With ODS folder corrected, the connection to the ODS and the database was not functional.

**Comment by skibum1869** [comment]
The adapter is not updating to the agent either.

logs

``` bash
hemsaw@hemipc:~/mtconnect$ sudo docker-compose logs
Attaching to mtc_agent, smartsaw_adapter, mosquitto, ods, watchtower
mosquitto           | 1712201275: mosquitto version 2.0.18 starting
mosquitto           | 1712201275: Config loaded from /mosquitto/config/mosquitto.conf.
mosquitto           | 1712201275: Opening ipv4 listen socket on port 1883.
mosquitto           | 1712201275: mosquitto version 2.0.18 running
mosquitto           | 1712201276: New connection from 10.0.0.1:42316 on port 1883.
mosquitto           | 1712201276: New bridge connected from 10.0.0.1:42316 as controlslab.bridge-01 (p2, c0, k60).
mosquitto           | 1712201276: New connection from 172.19.0.5:55028 on port 1883.
mosquitto           | 1712201276: New client connected from 172.19.0.5:55028 as mtc_agent (p2, c1, k10).
mtc_agent           | [2024-04-04 03:27:56.686784] [0x00007f39613da1f0] [debug]   Cannot file path: "/etc/mtconnect", No such file or directory
mtc_agent           | [2024-04-04 03:27:56.686923] [0x00007f39613da1f0] [debug]   Cannot file path: "/usr/local/etc/mtconnect", No such file or directory
mtc_agent           | [2024-04-04 03:27:56.686994] [0x00007f39613da1f0] [debug]   Cannot file path: "/usr/local/share/mtconnect", No such file or directory
mtc_agent           | MTConnect Agent Version 2.3.0.4 - built on Wed Mar 20 22:30:16 2024
mtc_agent           | [2024-04-04 03:27:56.687193] [0x00007f39613da1f0] [debug]   Found file '/mtconnect/config/agent.cfg'  in path "/home/agent"
mtc_agent           | [2024-04-04 03:27:56.687249] [0x00007f39613da1f0] [debug]   Loading configuration from: "/mtconnect/config/agent.cfg"
mtc_agent           | Loading configuration from:"/mtconnect/config/agent.cfg"
mtc_agent           | 2024-04-04T03:27:56.701728Z (0x00007f39613da1f0) [warning] MAIN->AgentConfiguration::initialize->AgentConfiguration::loadConfig->AgentConfiguration::loadSinks: The following path "/mtconnect/data/styles/old" is not a regular file: "/mtconnect/data/styles/old"
mtc_agent           | 2024-04-04T03:27:56.702204Z (0x00007f39613da1f0) [warning] MAIN->AgentConfiguration::initialize->AgentConfiguration::loadConfig->AgentConfiguration::loadSinks: The following path "/mtconnect/data/styles/old" is not a regular file: "/mtconnect/data/styles/old"
mtc_agent           | 2024-04-04T03:43:05.027840Z (0x00007f3960c99b30) [warning] SessionImpl::requested->Session::fail: Operation failed: Could not read request
mtc_agent           | 2024-04-04T03:43:05.027944Z (0x00007f3960c99b30) [warning] SessionImpl::requested->Session::fail: Closing: end of stream - end of stream
mtc_agent           | 2024-04-04T03:43:15.379591Z (0x00007f3960c76b30) [warning] SessionImpl::requested->Session::fail: Operation failed: Could not read request
mtc_agent           | 2024-04-04T03:43:15.379692Z (0x00007f3960c76b30) [warning] SessionImpl::requested->Session::fail: Closing: end of stream - end of stream
ods                 | MongoDB connection verification failed: No suitable servers found: `serverSelectionTimeoutMS` expired: [Failed to connect to target host: localhost:27017]: generic server error (Source: database->database->constructor)
ods                 | Listening on 0.0.0.0:9625 (Source: od_server->listener)
smartsaw_adapter    | Loading afg file
smartsaw_adapter    | Input file provided: /adapter/config/SmartSaw_DC_22_East.afg (Source: open_afg_file)
smartsaw_adapter    | Retrieved Product Name: NX1P2
smartsaw_adapter    | Communication protocol identified: CIP
smartsaw_adapter    | Protocol: CIP
smartsaw_adapter    | PLC Host: 10.0.0.2
smartsaw_adapter    | Server Host: 0.0.0.0
smartsaw_adapter    | Server Port: 9800
smartsaw_adapter    | in do_accept (Source: adapter->adapter_listener->do_accept)
smartsaw_adapter    | in do_accept (Source: adapter->adapter_listener->do_accept)
smartsaw_adapter    | in do_accept (Source: adapter->adapter_listener->do_accept)
smartsaw_adapter    | in do_accept (Source: adapter->adapter_listener->do_accept)
watchtower          | time="2024-04-04T03:27:56Z" level=info msg="Watchtower 1.7.1"
watchtower          | time="2024-04-04T03:27:56Z" level=info msg="Using no notifications"
watchtower          | time="2024-04-04T03:27:56Z" level=info msg="Checking all containers (except explicitly disabled with label)"
watchtower          | time="2024-04-04T03:27:56Z" level=info msg="Scheduling first run: 2024-04-05 03:27:56 +0000 UTC"
watchtower          | time="2024-04-04T03:27:56Z" level=info msg="Note that the first check will be performed in 23 hours, 59 minutes, 59 seconds"
```

**Comment by skibum1869** [comment]
does the mongo need to be run in another docker container so it doesn't need to be installed on the host os?

**Comment by emprarthanak** [comment]
> The change log must be updated to reflect the docker-compose changes for the daemon, changing the ODS and Adapter to docker containers.

Will update the change log

**Comment by emprarthanak** [comment]
> With ODS folder corrected, the connection to the ODS and the database was not functional.

As the MongoDB is running on host system, we manually need to start the MongoDB server for the ODS to connect.
Since, ODS docker container runs in the host network it should not have any issue with connecting to MongoDB running in the localhost

**Comment by emprarthanak** [comment]
> does the mongo need to be run in another docker container so it doesn't need to be installed on the host os?

We can run MongoDB as docker container, but we need to take care of persisting data and its pros and cons.

**Comment by emprarthanak** [comment]
> The change log must be updated to reflect the docker-compose changes for the daemon, changing the ODS and Adapter to docker containers.

We decided to give version 1.1.0.0 for docker-compose changes as the previous version was 1.0.1.0. Do you prefer any other version to be mentioned or is this ok ? Please, let us know

**Comment by skibum1869** [comment]
> We are planing to give version 1.1.0.0 for docker-compose changes as there are minor changes. Do you prefer any other version to be mentioned or is this ok ? Please, let us know

The version 1.1.0 is perfect. I agree that the changes are minor changes. 

> We can run MongoDB as docker container, but we need to take care of persisting data and its pros and cons.

We can link the volume to a local folder with a read-write status for persistence.

**Comment by skibum1869** [comment]
SmartSaw_Adapter is running, and we will monitor it for issues. We will test the ODS on Monday, as we have time to pull the machine out of production for a few hours.

**Comment by skibum1869** [comment]
I have updated the afg files to reflect a few of the commands.

**Comment by skibum1869** [comment]
Issue with the generate uuid on the afg.

``` bash
Loading afg file
smartsaw_adapter    | Input file provided: /adapter/config/SmartSaw_DC_22_East.afg (Source: open_afg_file)
smartsaw_adapter    | line 17:28 no viable alternative at input 'generate_uuid('
smartsaw_adapter    | trying_connection on 10.0.0.2:44818 (Source: tcp_client_session->try_connection)
smartsaw_adapter    | Entered wait_for_socket_available, socket_in_use_: false owns lock?: true (Source: tcp_client_session->wait_for_socket_available)
smartsaw_adapter    | Set socket_in_use_ to true (wasn't in use) (Source: tcp_client_session->wait_for_socket_available)
smartsaw_adapter    | Sending byte message (Source: tcp_client_session->send)
smartsaw_adapter    | Byte message sent (Source: tcp_client_session->send)
smartsaw_adapter    | Set socket_in_use_ to false (Source: tcp_client_session->set_socket_available)
smartsaw_adapter    | Entered wait_for_socket_available, socket_in_use_: false owns lock?: true (Source: tcp_client_session->wait_for_socket_available)
smartsaw_adapter    | Set socket_in_use_ to true (wasn't in use) (Source: tcp_client_session->wait_for_socket_available)
smartsaw_adapter    | Sending byte message (Source: tcp_client_session->send)
smartsaw_adapter    | Byte message sent (Source: tcp_client_session->send)
smartsaw_adapter    | Set socket_in_use_ to false (Source: tcp_client_session->set_socket_available)
smartsaw_adapter    | Entered wait_for_socket_available, socket_in_use_: false owns lock?: true (Source: tcp_client_session->wait_for_socket_available)
smartsaw_adapter    | Set socket_in_use_ to true (wasn't in use) (Source: tcp_client_session->wait_for_socket_available)
smartsaw_adapter    | Sending byte message (Source: tcp_client_session->send)
smartsaw_adapter    | Byte message sent (Source: tcp_client_session->send)
smartsaw_adapter    | Set socket_in_use_ to false (Source: tcp_client_session->set_socket_available)
smartsaw_adapter    | Retrieved Product Name: NX1P2
smartsaw_adapter    | Communication protocol identified: CIP
smartsaw_adapter    | Entered wait_for_socket_available, socket_in_use_: false owns lock?: true (Source: tcp_client_session->wait_for_socket_available)
smartsaw_adapter    | Set socket_in_use_ to true (wasn't in use) (Source: tcp_client_session->wait_for_socket_available)
smartsaw_adapter    | Set socket_in_use_ to false (Source: tcp_client_session->set_socket_available)
smartsaw_adapter    | client_session destructor called (Source: client_session)
smartsaw_adapter    | Protocol: CIP
smartsaw_adapter    | PLC Host: 10.0.0.2
smartsaw_adapter    | Server Host: 0.0.0.0
smartsaw_adapter    | Server Port: 9800
smartsaw_adapter    | in do_accept (Source: adapter->adapter_listener->do_accept)
smartsaw_adapter    | trying_connection on 10.0.0.2:44818 (Source: tcp_client_session->try_connection)
smartsaw_adapter    | client session connected_: true (Source: client_session->connection_loop)
```

``` 
cts_version: CTS2
platform: Sysmac
log_severity_level: info
plc_host: 10.0.0.2
server_host: 0.0.0.0
server_port: 9800


// ** Device: Saw
saw_avail =  "AVAILABLE"// Autogenerated from Agent
saw_functionalmode = gMTConnectData.SawFunctionalMode
saw_serial_number = gMachineInformation.Serial
cmd_adapter_version: true
cmd_mtconnect_version = comd "2.2"
cmd_shdr_version = comd "2.0"
cmd_uuid = cmd generate_uuid(manufacturer = "HEMSaw",model = "",serialNumber = gMachineInformation.Serial)
cmd_device_name = comd "Saw"
cmd_serial_number = comd  gMachineInformation.Serial // sends the serial number command
```

**Comment by skibum1869** [comment]
Generate uuid is working, and testing the ods

**Pull request merged**